pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -34,24 +35,22 @@ type bucketChunkReader struct {
 	// After chunks are loaded, mutex is no longer used.
 	mtx        sync.Mutex
 	stats      *queryStats
-	chunkBytes []*[]byte // Byte slice to return to the chunk pool on close.
+	chunkBytes *pool.BatchBytes
 }
 
-func newBucketChunkReader(ctx context.Context, block *bucketBlock) *bucketChunkReader {
+func newBucketChunkReader(ctx context.Context, block *bucketBlock, chunkBytes *pool.BatchBytes) *bucketChunkReader {
 	return &bucketChunkReader{
-		ctx:    ctx,
-		block:  block,
-		stats:  &queryStats{},
-		toLoad: make([][]loadIdx, len(block.chunkObjs)),
+		ctx:        ctx,
+		block:      block,
+		stats:      &queryStats{},
+		toLoad:     make([][]loadIdx, len(block.chunkObjs)),
+		chunkBytes: chunkBytes,
 	}
 }
 
 func (r *bucketChunkReader) Close() error {
 	r.block.pendingReaders.Done()
-
-	for _, b := range r.chunkBytes {
-		r.block.chunkPool.Put(b)
-	}
+	r.chunkBytes.Release()
```

Comment:
The `Release()` is already called by `BucketStore.Series()` in a `defer`. It's not a big deal because `Release()` can correctly handle being called twice, but from a design perspective looks weird to me. Generally speaking, I would expect that any resource passed to `newBucketChunkReader()` is released from outside, while any resource internally created by `newBucketChunkReader()` will be released by `bucketChunkReader.Close()`.

Reply:
I believe this is addressed by https://github.com/grafana/mimir/pull/3506/commits/8d5ef60029e0b13c365d2442978b642cf2e731cf